### PR TITLE
Implemented UDIV and SDIV Instructions

### PR DESCRIPTION
**Pull request by FernandoS27**

### CODE REVIEW

**Comment by FernandoS27** [comment]
meh, I'll fix it tommorow, too tired to fix the conflict now, (didn't have it before).

File: src/backend_x64/emit_x64.cpp [hunk 1]
```diff
@@ -1048,6 +1048,94 @@ void EmitX64<JST>::EmitMul64(EmitContext& ctx, IR::Inst* inst) {
...
+    auto args = ctx.reg_alloc.GetArgumentInfo(inst);
+
+    Xbyak::Reg64 rax = ctx.reg_alloc.ScratchGpr({HostLoc::RAX});
+    Xbyak::Reg64 rdx = ctx.reg_alloc.ScratchGpr({HostLoc::RDX});
+    Xbyak::Reg64 dividend = ctx.reg_alloc.UseGpr(args[0]);
+    Xbyak::Reg64 divisor = ctx.reg_alloc.UseGpr(args[1]);
+
+    Xbyak::Label end;
+
+    code->xor_(rax, rax);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: codesize: `code->xor_(code->eax, code->eax);`

File: src/backend_x64/emit_x64.cpp [hunk 2]
```diff
@@ -1048,6 +1048,94 @@ void EmitX64<JST>::EmitMul64(EmitContext& ctx, IR::Inst* inst) {
...
+    code->test(divisor, divisor);
+    code->jz(end);
+    code->mov(rax, dividend);
+    code->xor_(rdx,rdx);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: codesize: `code->xor_(code->edx, code->edx);`

File: src/backend_x64/emit_x64.cpp [hunk 3]
```diff
@@ -1048,6 +1048,94 @@ void EmitX64<JST>::EmitMul64(EmitContext& ctx, IR::Inst* inst) {
...
+    auto args = ctx.reg_alloc.GetArgumentInfo(inst);
+
+    Xbyak::Reg64 rax = ctx.reg_alloc.ScratchGpr({HostLoc::RAX});
+    ctx.reg_alloc.ScratchGpr({HostLoc::RDX});
+    Xbyak::Reg64 dividend = ctx.reg_alloc.UseGpr(args[0]);
+    Xbyak::Reg64 divisor = ctx.reg_alloc.UseGpr(args[1]);
+
+    Xbyak::Label end;
+
+    code->xor_(rax, rax);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64.cpp]: codesize: `code->xor_(code->eax, code->eax);`

**Comment by FernandoS27** [comment]
well it does not matter if zeroes all the register or not, at the end it just needs to be zero where it's needed.

I ran the tests with unicorn, you should add the ability to limit testing to certain instructions as the time it takes is considerable.

File: src/backend_x64/emit_x64_data_processing.cpp [hunk 4]
```diff
@@ -878,6 +878,90 @@ void EmitX64::EmitMul64(EmitContext& ctx, IR::Inst* inst) {
...
+    Xbyak::Reg64 dividend = ctx.reg_alloc.UseGpr(args[0]);
+    Xbyak::Reg64 divisor = ctx.reg_alloc.UseGpr(args[1]);
+
+    Xbyak::Label end;
+
+    code->xor_(rax, rax);
+    code->test(divisor, divisor);
+    code->jz(end);
+    code->mov(rax, dividend);
+    code->xor_(rdx,rdx);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64_data_processing.cpp]: Smaller codesize: `code->xor_(rdx.cvt32(), rdx.cvt32());`

File: src/backend_x64/emit_x64_data_processing.cpp [hunk 5]
```diff
@@ -878,6 +878,90 @@ void EmitX64::EmitMul64(EmitContext& ctx, IR::Inst* inst) {
...
+    auto args = ctx.reg_alloc.GetArgumentInfo(inst);
+
+    Xbyak::Reg64 rax = ctx.reg_alloc.ScratchGpr({HostLoc::RAX});
+    ctx.reg_alloc.ScratchGpr({HostLoc::RDX});
+    Xbyak::Reg64 dividend = ctx.reg_alloc.UseGpr(args[0]);
+    Xbyak::Reg64 divisor = ctx.reg_alloc.UseGpr(args[1]);
+
+    Xbyak::Label end;
+
+    code->xor_(rax, rax);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64_data_processing.cpp]: Smaller codesize: `code->xor_(rax.cvt32(), rax.cvt32());`

File: src/backend_x64/emit_x64_data_processing.cpp [hunk 6]
```diff
@@ -878,6 +878,90 @@ void EmitX64::EmitMul64(EmitContext& ctx, IR::Inst* inst) {
...
+    auto args = ctx.reg_alloc.GetArgumentInfo(inst);
+
+    Xbyak::Reg64 rax = ctx.reg_alloc.ScratchGpr({HostLoc::RAX});
+    Xbyak::Reg64 rdx = ctx.reg_alloc.ScratchGpr({HostLoc::RDX});
+    Xbyak::Reg64 dividend = ctx.reg_alloc.UseGpr(args[0]);
+    Xbyak::Reg64 divisor = ctx.reg_alloc.UseGpr(args[1]);
+
+    Xbyak::Label end;
+
+    code->xor_(rax, rax);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64_data_processing.cpp]: Smaller codesize: `code->xor_(rax.cvt32(), rax.cvt32());`

File: src/backend_x64/emit_x64_data_processing.cpp [hunk 7]
```diff
@@ -878,6 +878,90 @@ void EmitX64::EmitMul64(EmitContext& ctx, IR::Inst* inst) {
...
+    Xbyak::Reg32 dividend = ctx.reg_alloc.UseGpr(args[0]).cvt32();
+    Xbyak::Reg32 divisor = ctx.reg_alloc.UseGpr(args[1]).cvt32();
+
+    Xbyak::Label end;
+
+    code->xor_(eax, eax);
+    code->test(divisor, divisor);
+    code->jz(end);
+    code->mov(eax, dividend);
+    code->xor_(edx,edx);
```
Comment (MerryMage) [inline src/backend_x64/emit_x64_data_processing.cpp]: Formatting (missing space): `code->xor_(edx, edx);`

**Comment by FernandoS27** [comment]
Changes made.

**Pull request merged**